torch/distributed/rpc/api.py

```diff
@@ -172,38 +172,41 @@ def _wait_all():
             del _thread_local_var.future_list
 
 @_require_initialized
-def _all_gather(obj, timeout=UNSET_RPC_TIMEOUT):
+def _all_gather(obj, worker_names=None, timeout=UNSET_RPC_TIMEOUT):
     r"""
     This is similar to torch.distributed.all_gather(), but is using RPC. It
     picks the worker with the smallest name (alphabetic order) as the leader.
     Then all followers send their data ``obj`` to the leader. After the leader
     has received all, it will broadcast the results back to all followers. This
     function blocks until all workers have received the gathered results.
     """
-    assert (
-        _ALL_WORKER_NAMES is not None
-    ), "`_ALL_WORKER_NAMES` is not initialized for `def _all_gather`."
-    leader_name = sorted(_ALL_WORKER_NAMES)[0]
+    if not worker_names:
+        assert (
+            _ALL_WORKER_NAMES is not None
+        ), "`_ALL_WORKER_NAMES` is not initialized for `def _all_gather`."
+        worker_names = _ALL_WORKER_NAMES
+    leader_name = sorted(worker_names)[0]
 
     self_name = _get_current_rpc_agent().get_worker_info().name
 
-    global _all_gather_sequence_id
     with _all_gather_dict_lock:
-        sequence_id = _all_gather_sequence_id
-        _all_gather_sequence_id += 1
+        concat_names = "".join(sorted(worker_names))
+        sequence_num = _all_gather_sequence_id.get(concat_names, 0)
```

Comment:
I see that we properly clean up this key for `_all_gather_sequence_id` below when exiting the barrier so that future calls with the same set of worker_names will start at 0 as expected.

However, is it possible to still run into races here? If there are concurrent calls to `barrier()` (i.e. multiple threads) with the same set of workers:

Thread 1 creates the key in `_all_gather_sequence_id`
Thread 1 releases lock 
Thread 2 calls with the same set of arguments, so it sees an already incremented sequence_num, which could cause the tracking of states to be incorrect.

Reply:
The cleanup done at the end of barrier is for `_all_gather_sequence_id_to_states`. This var keeps tracks of the barrier state for a single sequence id. This cleanup is needed because there is an existing memory leak (small one) where each call to `_all_gather` will leave behind an `AllGatherStates` class in the global dictionary.

The `_all_gather_sequence_id` is an ID representing a particular barrier operation. It is strictly increasing so that concurrent calls will still work. I'll rewrite your example above to demonstrate what is happening:

- worker1, Thread 1 performs barrier on ["worker1", "worker2"], retrieve `_all_gather_sequence_id` (value 0), increment `_all_gather_sequence_id`
- worker1, Thread 1 release lock
- worker1, Thread 2 performs barrier on ["worker1", "worker2"], retrieve `_all_gather_sequence_id` (value 1), increment `_all_gather_sequence_id`
- Thread1 and Thread2 block waiting on the barrier from worker2
- When worker2 calls barrier ["worker1", "worker2"] the first time, it will allow worker1, Thread1 to make progress. The second call to barrier will allow worker1, Thread2 to make progress.

Reply:
Sounds good. Could we add a unittest to verify this multithreaded behavior?